polynote-server/src/main/scala/polynote/NotebookRunner.scala

```diff
@@ -0,0 +1,225 @@
+package polynote
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Path, Paths}
+import java.time.{Instant, LocalDate, OffsetDateTime, ZoneId, ZoneOffset}
+import java.time.format.{DateTimeFormatter, DateTimeFormatterBuilder}
+
+import polynote.app.MainArgs
+import polynote.kernel.{BaseEnv, CellEnv, GlobalEnv, Kernel, KernelStatusUpdate}
+import polynote.kernel.environment.{CurrentNotebook, NotebookUpdates, PublishResult, PublishStatus}
+import polynote.kernel.task.TaskManager
+import polynote.kernel.util.Publish
+import polynote.messages.{CellID, Notebook, NotebookCell}
+import polynote.server.AppEnv
+import polynote.server.repository.{NotebookContent, NotebookRepository}
+import polynote.server.repository.format.NotebookFormat
+import polynote.server.repository.fs.FileSystems
+import zio.blocking.{Blocking, effectBlocking}
+import zio.{Has, IO, Queue, RIO, Ref, Task, UIO, URIO, URLayer, ZIO, ZLayer}
+import ZIO.{effect, effectTotal}
+import zio.clock, clock.Clock
+import zio.interop.catz._
+import zio.stream.Take
+
+import scala.annotation.tailrec
+
+object NotebookRunner {
+
+  type CellBaseEnv = BaseEnv with GlobalEnv with CurrentNotebook with TaskManager with PublishStatus
+  type KernelFactoryEnv = CellBaseEnv with NotebookUpdates
+
+  def main: ZIO[AppEnv, String, Int] = for {
+    mainArgs      <- MainArgs.access
+    runnerArgs    <- Args.parseOrErr(mainArgs.rest)
+    statusUpdates <- Queue.unbounded[Take[Nothing, KernelStatusUpdate]]
+    publishers     = PublishStatus.layer(statusUpdates)
+    results       <- ZIO.foreach(runnerArgs.inputFiles)(path => runPath(runnerArgs, path))
+      .provideSomeLayer[AppEnv](publishers)
+      .orDie
+  } yield 0
+
+
+  def runPath(runnerArgs: Args, pathStr: String): RIO[AppEnv with PublishStatus, Unit] = for {
+    path   <- parsePath(pathStr)
+    result <- loadNotebook(path) >>= runNotebook
+    _      <- writeResult(runnerArgs, path, result)
+  } yield ()
+
+  def parsePath(path: String): RIO[Blocking, Path] = effect(Paths.get(path)).flatMap {
+    case path if path.startsWith(".") => effectBlocking(path.toAbsolutePath)
+    case path => ZIO.succeed(path)
+  }
+
+  def loadNotebook(path: Path): RIO[AppEnv, Notebook] = path match {
+    case path if path.isAbsolute =>
+      for {
+        format   <- NotebookFormat.getFormat(path)
+        content  <- FileSystems.readPathAsString(path)
+        notebook <- format.decodeNotebook(path.toString.stripSuffix(s".${format.extension}"), content)
+      } yield notebook
+
+    case path => NotebookRepository.loadNotebook(path.toString)
+  }
+
+  def runNotebook(notebook: Notebook): RIO[AppEnv with PublishStatus, Notebook] =
+    runWithEnv.provideSomeLayer[AppEnv with PublishStatus] {
+      CurrentNotebook.make(notebook) ++ TaskManager.layer ++ NotebookUpdates.empty
+    }
+
+  def runWithEnv: ZIO[AppEnv with CellBaseEnv with NotebookUpdates, Throwable, Notebook] =
```

Comment:
Will this load dependencies defined in the notebook?

Reply:
It just launches a kernel by whatever mechanism is configured, so I think the kernel ought to do that. No?

Reply:
🤦 again. 

You can tell I've been doing nothing but Python and Typescript for about a month... sigh. Don't mind me 😧